### PR TITLE
[StructuralMechanicsApplication] Refactor common methods to static utilities

### PR DESCRIPTION
**Pull request by loumalouomega**
This is a cherry-pick of the changes reverted in #5951. Moving some common methods to static utilities

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 1]
```diff
@@ -8,20 +8,166 @@
...
+#include "includes/checks.h"
 #include "structural_mechanics_element_utilities.h"
 #include "structural_mechanics_application_variables.h"
 #include "utilities/math_utils.h"
 
 namespace Kratos {
 namespace StructuralMechanicsElementUtilities {
 
+int BaseElementCheck(
+    const Element* pElement,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]:
```suggestion
    const Element& rElement,
```

From looking at the code you don't need the pointer

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 2]
```diff
@@ -8,20 +8,166 @@
...
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY)
+    KRATOS_CHECK_VARIABLE_KEY(VOLUME_ACCELERATION)
+    KRATOS_CHECK_VARIABLE_KEY(THICKNESS)
+
+    // Check that the element's nodes contain all required SolutionStepData and Degrees of freedom
+    for ( IndexType i = 0; i < number_of_nodes; i++ ) {
+        const NodeType &rnode = r_geometry[i];
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT,rnode)
+//         KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(VELOCITY,rnode)
+//         KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ACCELERATION,rnode)
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: please remove

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 3]
```diff
@@ -8,20 +8,166 @@
...
+    return 0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ComputeEquivalentF(
+    const Element* pElement,
+    Matrix& rF,
+    const Vector& rStrainTensor
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: I would first put all input then all output
I think this is an unwritten convention

=> switch the order of rF and rStrainTensor

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 4]
```diff
@@ -8,20 +8,166 @@
...
+            rB(5, i*3    ) = rDN_DX(i, 2);
+            rB(5, i*3 + 2) = rDN_DX(i, 0);
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+array_1d<double, 3> GetBodyForce(
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:80]: :+1:

I will hopefully use this too in the future :)

Do you think there is a solution for using this also for elements without integration-points (e.g. beam)?

**Review by philbucher — COMMENTED** [review]
overall looks nice to me, only minor comments
Thanks

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 5]
```diff
@@ -965,9 +908,29 @@ int  SmallDisplacementMixedVolumetricStrainElement::Check(const ProcessInfo& rCu
 {
     KRATOS_TRY
 
-    int base_element_check = SmallDisplacementMixedVolumetricStrainElement::BaseType::Check(rCurrentProcessInfo);
+    int check = SmallDisplacementMixedVolumetricStrainElement::BaseType::Check(rCurrentProcessInfo);
+
+    // Base check
+    check = StructuralMechanicsElementUtilities::BaseElementCheck(this, rCurrentProcessInfo);
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: In here you check for the dynamic variables. Although they are not used right now, I'd keep the check for when we implement the dynamics of the element.

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 6]
```diff
@@ -965,9 +908,29 @@ int  SmallDisplacementMixedVolumetricStrainElement::Check(const ProcessInfo& rCu
...
+
+    // Verify that the variables are correctly initialized
+    KRATOS_CHECK_VARIABLE_KEY(THICKNESS)
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: I think this is already done in the `BaseElementCheck` you have implemented.

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 7]
```diff
@@ -965,9 +908,29 @@ int  SmallDisplacementMixedVolumetricStrainElement::Check(const ProcessInfo& rCu
...
+    for ( IndexType i = 0; i < r_geometry.size(); i++ ) {
+        const NodeType& r_node = r_geometry[i];
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(VOLUMETRIC_STRAIN,r_node)
+
+        KRATOS_CHECK_DOF_IN_NODE(VOLUMETRIC_STRAIN, r_node)
+    }
+
+    // Check constitutive law
+    if ( mConstitutiveLawVector.size() > 0 ) {
+        return mConstitutiveLawVector[0]->Check( GetProperties(), GetGeometry(), rCurrentProcessInfo );
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: Maybe it is a good idea that we also put this in the `BaseElementCheck`.

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 8]
```diff
@@ -8,20 +8,166 @@
...
 // Project includes
+#include "includes/checks.h"
 #include "structural_mechanics_element_utilities.h"
 #include "structural_mechanics_application_variables.h"
 #include "utilities/math_utils.h"
 
 namespace Kratos {
 namespace StructuralMechanicsElementUtilities {
 
+int BaseElementCheck(
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: I found that `BaseElementCheck` is a misleading name as this is not the base class of the elements that use it. Something like `SolidElementCheck`, `BaseCheck` o `StandardCheck` sounds better to me.

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 9]
```diff
@@ -8,20 +8,166 @@
...
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void CalculateB(
+    const Element* pElement,
+    Matrix& rB,
+    const Matrix& rDN_DX
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: You can also take into account @philbucher 's comment above in here (switch `rDN_DX` and `rB`).

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h [hunk 10]
```diff
@@ -21,8 +24,78 @@
...
+typedef std::size_t IndexType;
+
+/// Node type definition
+typedef Node<3> NodeType;
+
+/// Geometry definitions
+typedef Geometry<NodeType> GeometryType;
+
+/// The zero tolerance
+static constexpr double tolerance = std::numeric_limits<double>::epsilon();
```
Comment (rubenzorrilla) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h]: I think this is not needed.

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 4 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp:80]: Yes, but should be updated first

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 5 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: The dynamic are only checked that are registered, even if the problem is static will not give any issue (would mean you are not importing the core :P)

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 6 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: Typo, should be VOLUMETRIC_STRAIN

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 7 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: Ok

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp (hunk 8 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: OK

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h (hunk 10 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.h]: Nope, copy paste

**Comment by loumalouomega** [comment]
I think I applied all your comments  @philbucher and @rubenzorrilla

**Comment by RiccardoRossi** [comment]
please template the methods in the matrix types

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 11]
```diff
@@ -8,20 +8,173 @@
...
+        rF(2,1) = 0.5*rStrainTensor(4);
+        rF(2,2) = 1.0+rStrainTensor(2);
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void CalculateB(
+    const Element& rElement,
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: all of these methods should be templated in the matrix types. this is important to also allow using them when the mattix type is static

**Review by RiccardoRossi — COMMENTED** [review]
templating

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp [hunk 12]
```diff
@@ -808,33 +767,17 @@ void SmallDisplacementMixedVolumetricStrainElement::CalculateEquivalentStrain(Ki
 /***********************************************************************************/
 
 void SmallDisplacementMixedVolumetricStrainElement::ComputeEquivalentF(
-    Matrix &rF,
-    const Vector &rStrainTensor) const
+    Matrix& rF,
+    const Vector& rStrainTensor
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp:771]: template in MatrixType and VectorType

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 13]
```diff
@@ -8,20 +8,173 @@
...
+
+    return 0;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void ComputeEquivalentF(
+    const Element& rElement,
+    const Vector& rStrainTensor,
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: template in MatrixType and VectorType

File: applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp [hunk 14]
```diff
@@ -8,20 +8,173 @@
...
+        N = r_geometry.ShapeFunctionsValues(N, rIntegrationPoints[PointNumber].Coordinates());
+        for (IndexType i_node = 0; i_node < r_geometry.size(); ++i_node)
+            noalias(body_force) += N[i_node] * density * r_geometry[i_node].FastGetSolutionStepValue(VOLUME_ACCELERATION);
+    }
+
+    return body_force;
+}
+
 bool ComputeLumpedMassMatrix(
     const Properties& rProperites,
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_utilities/structural_mechanics_element_utilities.cpp]: typo in "properties"

**Review by RiccardoRossi — COMMENTED** [review]
same

**Comment by loumalouomega** [comment]
Done @RiccardoRossi

**Comment by RiccardoRossi** [comment]
thx, this now looks good to me. I let either @philbucher or @rubenzorrilla to approve, but you have my green light

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp (hunk 5 above)
Comment (rubenzorrilla) [reply applications/StructuralMechanicsApplication/custom_elements/small_displacement_mixed_volumetric_strain_element.cpp]: Yeah!

**Review by rubenzorrilla — APPROVED** [review]
Fine for me!. Thanks for the implementation.

**Review by philbucher — APPROVED** [review]
thanks :+1:

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**